src/main/java/org/disq_bio/disq/HtsjdkReadsRddStorage.java

```diff
@@ -0,0 +1,280 @@
+   * control the format to write in (BAM/CRAM/SAM, if not clear from the path extension), and the
+   * number of files to write (single vs. multiple).
+   *
+   * @param htsjdkReadsRdd a {@link HtsjdkReadsRdd} containing the header and the reads
+   * @param path the file or directory to write to
+   * @param writeOptions options to control aspects of how to write the reads (e.g. {@link
+   *     FormatWriteOption} and {@link FileCardinalityWriteOption}
+   * @throws IOException if an IO error occurs while writing
+   */
+  public void write(HtsjdkReadsRdd htsjdkReadsRdd, String path, WriteOption... writeOptions)
+      throws IOException {
+    FormatWriteOption formatWriteOption = null;
+    FileCardinalityWriteOption fileCardinalityWriteOption = null;
+    TempPartsDirectoryWriteOption tempPartsDirectoryWriteOption = null;
+    for (WriteOption writeOption : writeOptions) {
+      if (writeOption instanceof FormatWriteOption) {
+        formatWriteOption = (FormatWriteOption) writeOption;
+      } else if (writeOption instanceof FileCardinalityWriteOption) {
+        fileCardinalityWriteOption = (FileCardinalityWriteOption) writeOption;
+      } else if (writeOption instanceof TempPartsDirectoryWriteOption) {
+        tempPartsDirectoryWriteOption = (TempPartsDirectoryWriteOption) writeOption;
+      }
+    }
+
+    if (formatWriteOption == null) {
+      formatWriteOption = inferFormatFromPath(path);
+    }
+
+    if (formatWriteOption == null) {
+      throw new IllegalArgumentException(
+          "Path does not end in BAM, CRAM, or SAM extension, and format not specified.");
+    }
+
+    if (fileCardinalityWriteOption == null) {
+      fileCardinalityWriteOption = inferCardinalityFromPath(path);
+    }
+
+    String tempPartsDirectory = null;
+    if (tempPartsDirectoryWriteOption != null) {
+      tempPartsDirectory = tempPartsDirectoryWriteOption.getTempPartsDirectory();
+    } else if (fileCardinalityWriteOption == FileCardinalityWriteOption.SINGLE) {
+      tempPartsDirectory = path + ".parts";
+    }
+
+    getSink(formatWriteOption, fileCardinalityWriteOption)
+        .save(
+            sparkContext,
+            htsjdkReadsRdd.getHeader(),
+            htsjdkReadsRdd.getReads(),
+            path,
+            referenceSourcePath,
+            tempPartsDirectory);
+  }
+
+  private FormatWriteOption inferFormatFromPath(String path) {
+    SamFormat samFormat = SamFormat.fromPath(path);
+    return samFormat == null ? null : samFormat.toFormatWriteOption();
+  }
+
+  private FileCardinalityWriteOption inferCardinalityFromPath(String path) {
+    SamFormat samFormat = SamFormat.fromPath(path);
+    return samFormat == null
+        ? FileCardinalityWriteOption.MULTIPLE
+        : FileCardinalityWriteOption.SINGLE;
+  }
+
+  private AbstractSamSink getSink(
+      FormatWriteOption formatWriteOption, FileCardinalityWriteOption fileCardinalityWriteOption) {
+    switch (fileCardinalityWriteOption) {
+      case SINGLE:
```

Comment:
Same as before - the WriteOptions can have a method to get the sink (both FormatWriteOption and FileCardinalityOption)

Reply:
Done